Add routes command

adleong opened the pull request:
Add a routes command which displays per-route stats for services that have service profiles defined.

```
$ bin/linkerd routes webapp --from deploy/traffic
               ROUTE   SUCCESS      RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99   TLS
/authors/{id}/delete     5.64%   7.4rps          37ms         247ms         695ms    0%
  /books/{id}/delete     5.75%   7.2rps          29ms         160ms         680ms    0%
              /books    44.64%   1.9rps          71ms        1300ms        1860ms    0%
         /books/{id}   100.00%   0.8rps          75ms         470ms         494ms    0%
     /book/{id}/edit    50.00%   0.8rps         138ms         265ms         293ms    0%
       /authors/{id}   100.00%   0.4rps          30ms          95ms          99ms    0%
            /authors   100.00%   0.4rps          30ms          90ms          98ms    0%
```

This change has three parts:
* A new public-api RPC called `TopRoutes` which serves per-route stat data about a service
* An implementation of TopRoutes in the public-api service.  This implementation reads per-route data from Prometheus.  This is very similar to how the StatSummaries RPC and much of the code was able to be refactored and shared.
* A new CLI command called `routes` which displays the per-route data in a tabular or json format.  This is very similar to the `stat` command and much of the code was able to be refactored and shared.

The easiest way to test this change is to do the following:
1. install Linkerd off this branch
1. inject and deploy the books app (https://run.linkerd.io/booksapp.yml)
1. apply this service profile: https://gist.github.com/adleong/1f2d1d79f93f621cdd64f852f720fa16
1. run `bin/linkerd routes webapp --from deploy/traffic`

Note that as of the currently targeted proxy version, only outbound route stats are supported so the `--from` flag must be included in order to see data.  This restriction will be lifted in an upcoming PR once we add support for inbound route stats as well.

adleong commented on cli/cmd/routes.go:
```diff
@@ -0,0 +1,281 @@
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the specified resource")
+	cmd.PersistentFlags().StringVarP(&options.timeWindow, "time-window", "t", options.timeWindow, "Stat window (for example: \"10s\", \"1m\", \"10m\", \"1h\")")
+	cmd.PersistentFlags().StringVar(&options.fromResource, "from", options.fromResource, "If present, restricts outbound stats from the specified resource name")
```
Note that the `--to` flag is not supported here because the target of a routes command must be a service.

siggy commented:
Related to CI failure, it looks like `top_routes.go` mentions `public.TopRoutesRequest_ToResource`, but that is not defined in `public.proto`, so running `bin/protoc-go.sh` changes `public.pb.go` and breaks the build.

```bash
$ go test -race ./...
# github.com/linkerd/linkerd2/controller/api/public
controller/api/public/top_routes.go:36:8: undefined: public.TopRoutesRequest_ToResource
controller/api/public/top_routes.go:37:21: undefined: public.TopRoutesRequest_ToResource
# github.com/linkerd/linkerd2/controller/api/public [github.com/linkerd/linkerd2/controller/api/public.test]
controller/api/public/top_routes.go:36:8: undefined: public.TopRoutesRequest_ToResource
controller/api/public/top_routes.go:37:21: undefined: public.TopRoutesRequest_ToResource
FAIL	github.com/linkerd/linkerd2/cli/cmd [build failed]
```

siggy commented on cli/cmd/routes.go:
```diff
@@ -0,0 +1,281 @@
+	return renderRouteStats(resp, options), nil
+}
+
+func renderRouteStats(resp *pb.TopRoutesResponse, options *routesOptions) string {
```
consider refactoring with `renderStats`

adleong commented:
@siggy I've incorporated your suggestions and added some tests.  PTAL.  I've also updated the description of this PR with instructions for how this can be tested.  I'm going to defer adding any documentation in the repo because this feature is incomplete and will need more comprehensive documentation once both inbound and outbound stats are supported.

siggy commented on controller/api/public/test_helper.go:
```diff
@@ -179,3 +184,72 @@ func GenStatSummaryResponse(resName, resType string, resNs []string, counts *Pod
+	return mockProm, fakeGrpcServer, nil
+}
+
+func (exp expectedStatRpc) verifyPromQueries(mockProm *MockProm) error {
```
👍

siggy commented on cli/cmd/root.go:
```diff
@@ -131,6 +132,43 @@ func validatedPublicAPIClient(retryDeadline time.Time) pb.ApiClient {
 	return hc.PublicAPIClient()
 }
 
+type statOptionsBase struct {
```
nice!

consider including `fromNamespace` and `fromResource` as part of `statOptionsBase`, that way you wouldn't need a separate `routesOptions` struct.

siggy commented on cli/cmd/test_helper.go:
```diff
@@ -47,3 +47,8 @@ func readOptionalTestFile(t *testing.T, fileName string) string {
 	return fileData
 }
+
+func diffCompareFile(t *testing.T, actual string, goldenFile string) {
```
👍
